File: app/src/main/java/ca/judacribz/gainzassist/util/firebase/Authentication.java
package ca.judacribz.gainzassist.util.firebase;

import android.app.Activity;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.widget.Toast;

import ca.judacribz.gainzassist.background.FirebaseService;
import ca.judacribz.gainzassist.models.WorkoutHelper;
import com.facebook.login.LoginManager;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.FirebaseNetworkException;
import com.google.firebase.auth.*;

import ca.judacribz.gainzassist.R;
import ca.judacribz.gainzassist.activities.authentication.Login;

import java.util.Arrays;


public class Authentication {

    // Constants
    // --------------------------------------------------------------------------------------------
    private static final FirebaseAuth mAuth = FirebaseAuth.getInstance();
    private static final int MIN_PASSWORD_LEN = 6;
    public static final int RC_SIGN_IN = 9001;
    // --------------------------------------------------------------------------------------------

    // Global Vars
    // --------------------------------------------------------------------------------------------
    private static boolean userCreated = false;
    // --------------------------------------------------------------------------------------------


    /* Wrapper function to create an user in Firebase */
    public static void createUser(final Activity act, String email, String password, final GoogleSignInClient signInClient) {
        mAuth.createUserWithEmailAndPassword(email, password)
                .addOnCompleteListener(act, new OnCompleteListener<AuthResult>() {

                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        handleOnComplete(act, signInClient, task);
                    }
                });
    }

    /* Wrapper function to log in an user using a firebase credential */
    public static void signIn(final Activity act, AuthCredential cred, final GoogleSignInClient signInClient) {
        mAuth.signInWithCredential(cred)
                .addOnCompleteListener(act, new OnCompleteListener<AuthResult>() {

                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        handleOnComplete(act, signInClient, task);
                    }
                });
    }

    private static void handleOnComplete(
            Activity act,
            GoogleSignInClient signInClient,
            @NonNull Task<AuthResult> task) {
        String msg;

        // Sign up success
        if (task.isSuccessful()) {
            userCreated = true;
            msg = act.getString(R.string.sign_up_success);

            // Sign up fail
        } else {
            userCreated = false;
            msg = getExceptionMsg(act, task.getException());
        }

        Toast.makeText(act, msg, Toast.LENGTH_SHORT).show();
        if (msg.equals(act.getString(R.string.txt_email_registered))) {
            ((Login) act).linkGoogle = true;
            googleSignIn(act, signInClient);
        }
    }

    public static void linkUser(final Activity act, AuthCredential cred, FirebaseUser currUser) {
        currUser.linkWithCredential(cred)
                .addOnCompleteListener(act, new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        if (task.isSuccessful()) {
                            Toast.makeText(act, "Linked!", Toast.LENGTH_SHORT).show();
                        } else {
                            Toast.makeText(act, "linkWithCredential:failure" + task.getException(), Toast.LENGTH_SHORT).show();
                        }
                    }
                });
    }


    public static void googleSignIn(Activity act, GoogleSignInClient signInClient) {
        Intent signInIntent = signInClient.getSignInIntent();
        act.startActivityForResult(signInIntent, RC_SIGN_IN);
    }

//    public static void facebookSignIn(Activity act) {
//        LoginManager.getInstance().logInWithReadPermissions(
//                act,
//                Arrays.asList("email", "public_profile")
//        );
//    }

    /* Wrapper function to sign out user */
    public static void signOut(Activity act, GoogleSignInClient signInClient) {
        FirebaseAuth.getInstance().signOut();

        signInClient.signOut().addOnCompleteListener(
                act,
                new OnCompleteListener<Void>() {
                    @Override
                    public void onComplete(@NonNull Task<Void> task) {
                    }
                }
        );

        act.stopService(new Intent(act, FirebaseService.class));
        (new WorkoutHelper(act)).deleteAllWorkouts();
    }

    /* Used by sign in and sign up to firebase to handle exception if they fail */
    private static String getExceptionMsg(Activity act, Exception taskEx) {
        String msg = "";

        if (taskEx != null) {
            try {
                throw taskEx;

            // No internet connection
            } catch (FirebaseNetworkException ex) {
                msg = act.getString(R.string.txt_network_needed);

            // Password does not meet minimum requirements
            } catch (FirebaseAuthWeakPasswordException ex) {
                msg = act.getString(R.string.err_invalid_password);

            // Email is already registered
            } catch (FirebaseAuthUserCollisionException ex) {
                msg = act.getString(R.string.txt_email_registered);

            // Email format is invalid
            } catch (FirebaseAuthInvalidCredentialsException ex) {
                msg = act.getString(R.string.err_invalid_email);

            // Other exceptions
            } catch (Exception ex) {
                msg = ex.toString();
            }
        }

        return msg;
    }

//    /* Validates login and sign up forms using email and password combination */
//    public static boolean validateForm(Activity activity, EditText txtEmail, EditText txtPassword) {
//        boolean formIsValid = false;
//        String email = txtEmail.getText().toString().trim();
//        String password = txtPassword.getText().toString().trim();
//
//        if (!email.isEmpty() && !password.isEmpty()) {
//            formIsValid = true;
//        } else {
//            if (email.isEmpty()) {
//                txtEmail.setError(activity.getString(R.string.err_required));
//            }
//
//            if (password.isEmpty()) {
//                txtPassword.setError(activity.getString(R.string.err_required));
//            }
//        }
//
//        if (formIsValid) {
//            // Check to see email is in the correct format
//            if (!Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
//                txtEmail.setError(activity.getString(R.string.err_required_email_format));
//                formIsValid = false;
//            }
//
//            // Check for password at min length
//            if (password.length() < MIN_PASSWORD_LEN) {
//                txtPassword.setError(activity.getString(R.string.err_required_password_min));
//                formIsValid = false;
//            }
//        }
//
//        return formIsValid;
//    }
}

File: app/src/main/java/ca/judacribz/gainzassist/util/firebase/Authentication.kt
package ca.judacribz.gainzassist.util.firebase

import android.app.Activity
import android.content.Intent
import android.util.Log
import android.widget.Toast
import ca.judacribz.gainzassist.R
import ca.judacribz.gainzassist.activities.authentication.Login
import ca.judacribz.gainzassist.background.FirebaseService
import ca.judacribz.gainzassist.models.db.WorkoutViewModel
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.firebase.FirebaseNetworkException
import com.google.firebase.auth.*
import android.arch.lifecycle.ViewModelProviders
import android.support.v4.app.FragmentActivity

object Authentication {

    private val mAuth = FirebaseAuth.getInstance()
    const val RC_SIGN_IN = 9001
    private var userCreated = false

    @JvmStatic
    val currentUser: FirebaseUser?
        get() = mAuth.currentUser

    @JvmStatic
    fun createUser(act: Activity, email: String, password: String, signInClient: GoogleSignInClient) {
        mAuth.createUserWithEmailAndPassword(email, password)
            .addOnCompleteListener(act) { task -> handleOnComplete(act, signInClient, task) }
    }

    @JvmStatic
    fun signIn(act: Activity, cred: AuthCredential, signInClient: GoogleSignInClient) {
        mAuth.signInWithCredential(cred)
            .addOnCompleteListener(act) { task -> handleOnComplete(act, signInClient, task) }
    }

    private fun handleOnComplete(
        act: Activity,
        signInClient: GoogleSignInClient,
        task: com.google.android.gms.tasks.Task<AuthResult>
    ) {
        val msg: String
        if (task.isSuccessful) {
            userCreated = true
            msg = act.getString(R.string.sign_up_success)
        } else {
            userCreated = false
            val ex = task.exception
            if (ex != null) {
                Log.e(Authentication::class.java.simpleName, "Firebase Auth Error: " + ex.javaClass.name)
                Log.e(Authentication::class.java.simpleName, "Message: " + ex.message)
            }
            msg = getExceptionMsg(act, ex)
            (act as Login).loginFail()
            if (msg == act.getString(R.string.txt_email_registered)) {
                act.linkGoogle = true
                act.googleLogin()
            }
        }
        Toast.makeText(act, msg, Toast.LENGTH_SHORT).show()
    }

    @JvmStatic
    fun linkUser(act: Activity, cred: AuthCredential, currUser: FirebaseUser) {
        currUser.linkWithCredential(cred)
            .addOnCompleteListener(act) { task ->
                if (task.isSuccessful) {
                    Toast.makeText(act, "Linked!", Toast.LENGTH_SHORT).show()
                } else {
                    Toast.makeText(act, "linkWithCredential:failure" + task.exception, Toast.LENGTH_SHORT).show()
                }
            }
    }

    @JvmStatic
    fun signOut(act: Activity, signInClient: GoogleSignInClient) {
        FirebaseAuth.getInstance().signOut()
        signInClient.signOut().addOnCompleteListener(act) {}
        act.stopService(Intent(act, FirebaseService::class.java))
        if (act is FragmentActivity) {
            ViewModelProviders.of(act).get(WorkoutViewModel::class.java).deleteAllWorkouts()
        }
    }

    private fun getExceptionMsg(act: Activity, taskEx: Exception?): String {
        var msg = ""
        if (taskEx != null) {
            try {
                throw taskEx
            } catch (ex: FirebaseNetworkException) {
                msg = act.getString(R.string.txt_network_needed)
            } catch (ex: FirebaseAuthWeakPasswordException) {
                msg = act.getString(R.string.err_invalid_password)
            } catch (ex: FirebaseAuthUserCollisionException) {
                msg = act.getString(R.string.txt_email_registered)
            } catch (ex: FirebaseAuthInvalidCredentialsException) {
                msg = act.getString(R.string.err_invalid_email)
            } catch (ex: Exception) {
                msg = "Login failed. Check Firebase configuration."
            }
        }
        return msg
    }
}
